### PR TITLE
Now each Colors.foo constant is-a Color and a color swatch

### LINKED ISSUE #6706: Colors.______ API usability issues
We saw this slight bit of friction in two UX studies.

`Colors.blue` is actually a map of blues. We saw two different participants type/code-complete `Colors.blue` and then be surprised that they then need to use the map syntax to find the right actual blue.

There's no code completion of values inside the map.

A few observations:

* Colors.blue is singular, but it's returning a collection, which is plural
* Code completion doesn't help after Colors.blue

Strawman suggestions:

* Colors.blues (make it plural)
* Enable code completion after "blues", so the user can see all the options (use an enum?)

Similar to how our icons work, code completion Icons._____ shows each option (with a pretty picture, if using Atom! :)

### PR DESCRIPTION
**Pull request by HansMuller**
Fixes: https://github.com/flutter/flutter/issues/6706

### CODE REVIEW

File: packages/flutter/lib/src/material/colors.dart [hunk 1]
```diff
@@ -2,7 +2,76 @@
...
 // found in the LICENSE file.
 
-import 'dart:ui' show Color;
+import 'dart:ui' show Color, hashValues;
+
+class _ColorSwatch extends Color {
+  const _ColorSwatch(int primary, this.swatch) : super(primary);
+
+  /// The shades of this color defined by material design.
+  final Map<int, Color> swatch;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: I would rather we didn't expose this. Why can't the places that take a map today just take a ColorSwatch instead?

File: packages/flutter/lib/src/material/colors.dart [hunk 2]
```diff
@@ -2,7 +2,76 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-import 'dart:ui' show Color;
+import 'dart:ui' show Color, hashValues;
+
+class _ColorSwatch extends Color {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: It doesn't make sense for this class to be private since it's the superclass of a public class.

File: packages/flutter/lib/src/material/colors.dart [hunk 3]
```diff
@@ -2,7 +2,76 @@
...
+  final Map<int, Color> swatch;
+
+  /// Returns an element of the [swatch] table.
+  Color operator [](int index) => swatch[index];
+
+  @override
+  bool operator ==(dynamic other) {
+    if (identical(this, other))
+      return true;
+    if (other is! _ColorSwatch)
```
Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: compare runtimeType rather than using `is`, otherwise it's possible a subclass will be `==` in one direction but not in teh other (yes we get this wrong all over the place)

File: packages/flutter/lib/src/material/colors.dart [hunk 4]
```diff
@@ -2,7 +2,76 @@
...
+/// Defines a single color as well a color swatch with ten shades of the color.
+///
+/// The color's shades are referred to by index. The greater the index, the
+/// darker the color. There are 10 valid indices: 50, 100, 200, ..., 900.
+/// The value of this color is the same the value of index 400, or [shade400].
+///
+/// See also:
+///
+///  * [Colors] which defines all of the standard material colors.
+class MaterialColor extends _ColorSwatch {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/colors.dart]: _ColorSwatch is what MaterialColor and MaterialAccentColor have in common.

File: packages/flutter/lib/src/material/colors.dart [hunk 5]
```diff
@@ -2,7 +2,76 @@
...
+  @override
+  String toString() => '$runtimeType(primary value: ${super.toString()})';
+
+}
+
+/// Defines a single color as well a color swatch with ten shades of the color.
+///
+/// The color's shades are referred to by index. The greater the index, the
+/// darker the color. There are 10 valid indices: 50, 100, 200, ..., 900.
+/// The value of this color is the same the value of index 400, or [shade400].
```
Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: why 400? you should probably link to the relevant place that says that 400 is the right default

File: packages/flutter/lib/src/material/colors.dart (hunk 4 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: oh i see, the accent class is separate.

File: packages/flutter/lib/src/material/colors.dart [hunk 6]
```diff
@@ -2,7 +2,76 @@
...
+/// accent color.
+///
+/// The color's shades are referred to by index, the colors with smaller
+/// indices are lighter, larger indices are darker. There are four valid
+/// indices: 100, 200, 400, and 700. The value of this color is the same
+/// the value of index 200, or [shade200].
+///
+/// See also:
+///
+///  * [Colors] which defines all of the standard material colors.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: comma before "which"

Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: (also above)

File: packages/flutter/lib/src/material/colors.dart [hunk 7]
```diff
@@ -19,6 +88,14 @@ import 'dart:ui' show Color;
 /// Colors.green[400]  // Selects a mid-range green.
 /// ```
 ///
+/// Each color constant is-a color and can used directly. For example
```
Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: `s/is-a color/is a [Color]/`

File: packages/flutter/lib/src/material/colors.dart (hunk 1 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/colors.dart]: Mostly they could although it's a more disruptive change. The most difficult case I ran into was FlutterLogoDecoration, since it really doesn't want to depend on material.dart

**Comment by Hixie** [comment]
Can you confirm that Colors.foo.shadeNNN is const?

e.g. does this work:

`const Color a = Colors.blue.shade400;`

File: packages/flutter/lib/src/material/colors.dart (hunk 5 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/colors.dart]: RATS. I was under the mistaken impression that 400 was gospel  However.  https://material.io/guidelines/style/color.html# says:

> Google suggests using the 500 colors as the primary colors in your app 

So I will use 500.

**Comment by Hixie** [comment]
Filed the const thing as https://github.com/dart-lang/sdk/issues/29098

File: packages/flutter/lib/src/material/colors.dart [hunk 8]
```diff
@@ -2,7 +2,83 @@
...
+
+/// A color that has a small table of related colors called a "swatch".
+///
+/// See also:
+///
+///  * [MaterialColor] and [MaterialAccentColor], which define material design
+///    primary and accent color swatches.
+///  * [Colors], which defines all of the standard material design colors.
+class ColorSwatch extends Color {
+  const ColorSwatch(int primary, this.swatch) : super(primary);
```
Comment (abarth) [inline packages/flutter/lib/src/material/colors.dart]: Please doc the constructor too.  We usually restate the class docs briefly.  For example,

```Creates a color that has a small table of related colors called a "swatch".```

File: examples/flutter_gallery/lib/demo/colors_demo.dart [hunk 9]
```diff
@@ -71,29 +71,33 @@ class ColorItem extends StatelessWidget {
   }
 }
 
-class ColorSwatchTabView extends StatelessWidget {
-  ColorSwatchTabView({ Key key, this.swatch }) : super(key: key) {
-    assert(swatch != null && swatch.isValid);
+class PaletteTabView extends StatelessWidget {
+  static const List<int> primaryKeys =  const <int>[50, 100, 200, 300, 400, 500, 600, 700, 800, 900];
```
Comment (Hixie) [inline examples/flutter_gallery/lib/demo/colors_demo.dart]: nit: extraneous space after =

File: examples/layers/widgets/gestures.dart [hunk 10]
```diff
@@ -106,28 +106,42 @@ class _GestureDemoState extends State<GestureDemo> {
...
-        case Colors.lime:       _swatch = Colors.yellow; break;
-        case Colors.yellow:     _swatch = Colors.amber; break;
-        case Colors.amber:      _swatch = Colors.orange; break;
-        case Colors.orange:     _swatch = Colors.deepOrange; break;
-        case Colors.deepOrange: _swatch = Colors.brown; break;
-        case Colors.brown:      _swatch = Colors.grey; break;
-        case Colors.grey:       _swatch = Colors.blueGrey; break;
-        default:                assert(false);
-      }
+      if (_swatch == Colors.blueGrey)
```
Comment (Hixie) [inline examples/layers/widgets/gestures.dart]: I'm surprised the switch no longer worked.

File: packages/flutter/lib/src/material/colors.dart [hunk 11]
```diff
@@ -2,7 +2,84 @@
...
+///
+///  * [MaterialColor] and [MaterialAccentColor], which define material design
+///    primary and accent color swatches.
+///  * [Colors], which defines all of the standard material design colors.
+class ColorSwatch extends Color {
+  // Creates a color that has a small table of related colors called a "swatch".
+  const ColorSwatch(int primary, this.swatch) : super(primary);
+
+  /// The palette of colors based on this one.
+  final Map<int, Color> swatch;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: I still think we should make this private. :-)

File: packages/flutter/lib/src/material/colors.dart (hunk 5 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: looks like you didn't update the dartdoc here.

Also, I think we should rephrase this to say that the value of the color _should_ be the same as the value of index 500, because nothing actually ensures it, so someone could easily create a version that wasn't like that.

Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: Also, please add that link to the See Also section

File: packages/flutter/lib/src/material/colors.dart [hunk 12]
```diff
@@ -2,7 +2,84 @@
...
+  Color get shade900 => swatch[900];
+}
+
+/// Defines a single accent color as well a swatch of four shades of the
+/// accent color.
+///
+/// The color's shades are referred to by index, the colors with smaller
+/// indices are lighter, larger indices are darker. There are four valid
+/// indices: 100, 200, 400, and 700. The value of this color is the same
+/// the value of index 200, or [shade200].
```
Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: same here with respect to "the value should be" not "is"

File: packages/flutter/lib/src/material/colors.dart [hunk 13]
```diff
@@ -2,7 +2,84 @@
...
+///  * [Colors], which defines all of the standard material colors.
+class MaterialAccentColor extends ColorSwatch {
+  const MaterialAccentColor(int primary, Map<int, Color> swatch) : super(primary, swatch);
+  Color get shade100 => swatch[100];
+  Color get shade200 => swatch[200];
+  Color get shade400 => swatch[400];
+  Color get shade700 => swatch[700];
+}
 
 /// [Color] constants which represent Material design's
```
Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: they're no longer Color constants, they're ColorSwatch constants.

File: packages/flutter/lib/src/material/colors.dart [hunk 14]
```diff
@@ -478,20 +619,24 @@ class Colors {
...
+  static const int _cyanPrimaryValue = 0xFF26C6DA;
+  static const MaterialColor cyan = const MaterialColor(
+    _cyanPrimaryValue,
+    const <int, Color>{
+       50: const Color(0xFFE0F7FA),
+      100: const Color(0xFFB2EBF2),
+      200: const Color(0xFF80DEEA),
+      300: const Color(0xFF4DD0E1),
+      400: const Color(_cyanPrimaryValue),
+      500: const Color(_cyanPrimaryValue),
```
Comment (Hixie) [inline packages/flutter/lib/src/material/colors.dart]: !

File: packages/flutter/lib/src/painting/flutter_logo.dart [hunk 15]
```diff
@@ -31,47 +31,43 @@ enum FlutterLogoStyle {
...
-  _lightShade: const Color(0xFF42A5F5),
-  _darkShade: const Color(0xFF0D47A1)
-};
-
 /// An immutable description of how to paint Flutter's logo.
 class FlutterLogoDecoration extends Decoration {
   /// Creates a decoration that knows how to paint Flutter's logo.
   ///
-  /// The [swatch] controls the color used for the logo. The [style] controls
+  /// The [colors] control the colors used for the logo. The [style] controls
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/flutter_logo.dart]: `[colors]` here won't reference anything.

File: packages/flutter/lib/src/painting/flutter_logo.dart [hunk 16]
```diff
@@ -31,47 +31,43 @@ enum FlutterLogoStyle {
...
   /// whether and where to draw the "Flutter" label. If one is shown, the
   /// [textColor] controls the color of the label.
   ///
-  /// The [swatch], [textColor], and [style] arguments must not be null.
+  /// The [colors], [textColor], and [style] arguments must not be null.
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/flutter_logo.dart]: ditto

File: packages/flutter/lib/src/painting/flutter_logo.dart [hunk 17]
```diff
@@ -31,47 +31,43 @@ enum FlutterLogoStyle {
...
   }) : _position = style == FlutterLogoStyle.markOnly ? 0.0 : style == FlutterLogoStyle.horizontal ? 1.0 : -1.0, // ignore: CONST_EVAL_TYPE_BOOL_NUM_STRING
        // (see https://github.com/dart-lang/sdk/issues/26980 for details about that ignore statement)
        _opacity = 1.0;
 
-  FlutterLogoDecoration._(this.swatch, this.textColor, this.style, this._position, this._opacity, this.margin);
+  FlutterLogoDecoration._(this.lightColor, this.darkColor, this.textColor, this.style, this._position, this._opacity, this.margin);
 
-  /// The colors to use to paint the logo. This map should contain at least two
-  /// values, one for 400 and one for 900.
+  /// The logo graphic is painted with [lightColor] and [darkColor].
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/flutter_logo.dart]: add a one-line description of lightColor, e.g.:

```
/// The lighter of the two colors used to paint the logo.
```

File: packages/flutter/lib/src/painting/flutter_logo.dart [hunk 18]
```diff
@@ -31,47 +31,43 @@ enum FlutterLogoStyle {
...
+  /// are Flutter's secondary colors.
   ///
   /// In extreme cases where none of those four color schemes will work,
-  /// [Colors.pink], [Colors.purple], or [Colors.cyan] swatches can be used.
+  /// [Colors.pink], [Colors.purple], or [Colors.cyan] can be used.
   /// These are Flutter's tertiary colors.
-  final Map<int, Color> swatch;
+  final Color lightColor;
+
+  /// See [lightColor].
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/flutter_logo.dart]: add something like:

```
/// The darker of the two colors used to paint the logo.
```

...before a blank line then `See [lightColor] for a discussion of which colors are reasonable` or some such.

File: packages/flutter/lib/src/painting/flutter_logo.dart [hunk 19]
```diff
@@ -221,9 +210,9 @@ class FlutterLogoDecoration extends Decoration {
   @override
   String toString([String prefix = '', String prefixIndent ]) {
     final String extra = _inTransition ? ', transition $_position:$_opacity' : '';
-    if (swatch == null)
+    if (lightColor == null || darkColor == null)
       return '$prefix$runtimeType(null, $style$extra)';
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/flutter_logo.dart]: you can remove this if and its branch, it doesn't matter any more (it was just trying to prevent a null deref before)

File: examples/layers/widgets/gestures.dart (hunk 10 above)
Comment (HansMuller) [inline examples/layers/widgets/gestures.dart]: Dart doesn't allow switching on a type that overloads operator==.

File: packages/flutter/lib/src/material/colors.dart (hunk 14 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/colors.dart]: Oops. Good catch!

**Comment by Hixie** [comment]
Test? Things that could be tested are that all the colors in Colors.primaries have 500 as their primary and no duplicates (which would catch the cyan copy/pasta in this patch), or that FlutterLogoDecoration uses the right colors provided to the FlutterLogo widget, and no others (you could use the mock_canvas `paints` matcher for that), or that a MaterialColor created with 10 unique colors returns the same colors from `[NNN]` as from `swatchNNN`.

The goal is primarily to make sure we keep our coverage numbers up. colors.dart today has zero coverage.

File: examples/layers/widgets/gestures.dart (hunk 10 above)
Comment (Hixie) [inline examples/layers/widgets/gestures.dart]: huh. i had no idea.

**Comment by Hixie** [comment]
LGTM
